lib/pbench/client/__init__.py

```diff
@@ -0,0 +1,289 @@
+from typing import Dict, List, Optional, Union
+from urllib import parse
+
+import requests
+
+
+# A type defined to conform to the semantic definition of a JSON structure
+# with Python syntax.
+JSONSTRING = str
+JSONNUMBER = Union[int, float]
+JSONVALUE = Union["JSONOBJECT", "JSONARRAY", JSONSTRING, JSONNUMBER, bool, None]
+JSONARRAY = List[JSONVALUE]
+JSONOBJECT = Dict[JSONSTRING, JSONVALUE]
+JSON = JSONVALUE
+
+
+class PbenchClientError(Exception):
+    """
+    Base class for exceptions reported by the Pbench Server client.
+    """
+
+    pass
+
+
+class IncorrectParameterCount(PbenchClientError):
+    def __init__(self, api: str, cnt: int, uri_params: Optional[JSONOBJECT]):
+        self.api = api
+        self.cnt = cnt
+        self.uri_params = uri_params
+
+    def __str__(self) -> str:
+        return f"API template {self.api} requires {self.cnt} parameters ({self.uri_params})"
+
+
+class Pbench:
+    def __init__(self, host: str):
+        """
+        Create a Pbench Server client object. This validates that the server is
+        reachable, and responds to a GET on /api/v1/endpoints.
+
+        Args:
+            host    Pbench Server hostname
+        """
+        self.host = host
+        self.username = None
+        self.auth_token = None
+        self.session = None
+        self.connect()
+
+    def _headers(self, user_headers: Optional[Dict[str, str]] = None) -> Dict[str, str]:
+        """
+        Create an HTTP request headers dictionary. We always default to
+        "application/json" but each operation allows adding or overriding. If
+        we have an authentication token, then we also assume that we want to
+        use it.
+
+        Args:
+            user_headers:   Addition request headers
+
+        Returns:
+            Header dictionary
+        """
+        headers = user_headers if user_headers else {}
+        if self.auth_token:
+            headers["authorization"] = f"Bearer {self.auth_token}"
```

Comment:
> What I ought to do is use requests' `CaseInsensitiveDict` so I can reliably check presence

Doesn't the `in` operator work regardless of what the flavor of the dictionary is?  But, yeah, using the proper subtype for the default value (and in the callers) would be good.

Reply:
Right, _if_ the input is a plain `Dict` and if one side of the comparison is not lower case, then we end up with a problem.  I hadn't considered that both of those would be true at the same time.  So, yes, using `CaseInsensitiveDict` would be a superior solution (and, it might be worth checking here that that is the type that we actually received, since we sort of depend on it...).

Reply:
> it might be worth checking here that that is the type that we actually received, since we sort of depend on it.

I thought of that, but decided to *not* depend on it, by casting the `dict` we get into a `CaseInsensitiveDict` internally.

(Arguably it would have been "cleaner" to create an empty `CaseSensitiveDict` if the caller doesn't supply headers, but I decided not to bother since it doesn't matter.)

Reply:
> it doesn't matter

It doesn't matter because the resulting `dict` gets passed into the `requests` HTTP method function which treats all the keys as case-insensitive, and the only key we add is already lowercase?

So, the only reason why _we_ might want to use `CaseInsensitiveDict` is to ensure that we _don't_ supersede the `authorization` header when the caller supplies it?  That seems like a big hammer...but, I'm not immediately finding a smaller one that works....